cores/esp8266/Schedule.cpp

```diff
@@ -76,90 +77,99 @@ bool schedule_function_us(std::function<bool(void)>&& fn, uint32_t repeat_us, sc
+    esp8266::polledTimeout::periodicFastMs yieldNow(100); // yield every 100ms
+
+    while (sFirst)
+    {
+        sFirst->mFunc();
+
+        {
+            InterruptLock lockAllInterruptsInThisScope;
+
+            auto to_recycle = sFirst;
+            sFirst = sFirst->mNext;
+            if (!sFirst)
+                sLast = nullptr;
+            recycle_fn_unsafe(to_recycle);
+        }
+
+        if (yieldNow)
+        {
+            // because scheduled function are allowed to last:
+            // this is yield() in cont stack:
+            esp_schedule();
+            cont_yield(g_pcont);
+        }
+    }
 }
 
-void run_scheduled_functions(schedule_e policy)
+void run_scheduled_recurrent_functions ()
 {
     // Note to the reader:
     // There is no exposed API to remove a scheduled function:
     // Scheduled functions are removed only from this function, and
     // its purpose is that it is never called from an interrupt
     // (always on cont stack).
 
+    if (!rFirst)
+        return;
+
     static bool fence = false;
     {
         InterruptLock lockAllInterruptsInThisScope;
         if (fence)
             // prevent recursive calls from yield()
+            // (even if they are not allowed)
             return;
         fence = true;
     }
 
-    esp8266::polledTimeout::periodicFastMs yieldNow(100); // yield every 100ms
-    scheduled_fn_t* lastRecurring = nullptr;
-    scheduled_fn_t* nextCall = sFirst;
-    while (nextCall)
-    {
-        scheduled_fn_t* toCall = nextCall;
-        nextCall = nextCall->mNext;
-
-        // run scheduled function:
-        // - when its schedule policy allows it anytime
-        // - or if we are called at loop() time
-        // and
-        // - its time policy allows it
-        if (   (   toCall->policy == SCHEDULED_FUNCTION_WITHOUT_YIELDELAYCALLS
-                || policy == SCHEDULED_FUNCTION_ONCE_PER_LOOP)
-            && toCall->callNow)
-        {
-            if (toCall->mFunc())
-            {
-                // function stays in list
-                lastRecurring = toCall;
-            }
-            else
-            {
-                // function removed from list
-                InterruptLock lockAllInterruptsInThisScope;
-
-                if (sFirst == toCall)
-                    sFirst = sFirst->mNext;
-                else if (lastRecurring)
-                    lastRecurring->mNext = toCall->mNext;
+    recurrent_fn_t* prev = nullptr;
+    recurrent_fn_t* current = rFirst;
 
-                if (sLast == toCall)
-                    sLast = lastRecurring;
+    while (current)
+        if (current->callNow && !current->mFunc())
+        {
+            // remove function from stack
+            InterruptLock lockAllInterruptsInThisScope;
 
-                recycle_fn_unsafe(toCall);
-            }
+            auto to_ditch = current;
+            if (prev)
+                prev->mNext = current = current->mNext;
```

Comment:
Valid C, but maybe split this into two separate assignments so as to be explicit?